### PR TITLE
Rendition select

### PR DESCRIPTION
**Pull request by luwes**
Related #387

test urls: 

- https://media-chrome-git-fork-luwes-rendition-select-mux.vercel.app/examples/vanilla/control-elements/media-rendition-listbox.html
- https://media-chrome-git-fork-luwes-rendition-select-mux.vercel.app/examples/vanilla/control-elements/media-rendition-selectmenu.html

### CODE REVIEW

File: src/js/controller.js [hunk 1]
```diff
@@ -769,4 +811,14 @@ export const MediaUIRequestHandlers = {
...
     media.currentTime = seekable.end(seekable.length - 1);
   },
+  MEDIA_RENDITION_REQUEST: (media, event) => {
+    if (!media) return;
+
+    const renditionId = event.detail;
+    const selected = media.videoTracks?.[media.videoTracks?.selectedIndex ?? 0];
+
+    for (const rendition of selected?.renditions ?? []) {
+      rendition.enabled = renditionId === '' || rendition.id == renditionId;
```
Comment (heff) [inline src/js/controller.js:821]: Requesting a rendition of _empty string_ feels odd. Could we make "auto" a real value?

File: src/js/experimental/media-rendition-button.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * @attr {string} mediarenditionactive - (read-only) Set to the current active rendition.
+ *
+ * @cssproperty [--media-rendition-button-display = inline-flex] - `display` property of button.
+ */
+class MediaRenditionButton extends MediaChromeButton {
+  static get observedAttributes() {
+    return [
+      ...super.observedAttributes,
+      MediaUIAttributes.MEDIA_RENDITION_ENABLED,
+      MediaUIAttributes.MEDIA_RENDITION_ACTIVE,
```
Comment (heff) [inline src/js/experimental/media-rendition-button.js:48]: This is kind of showing my point from above, that the UI doesn't care about `MEDIA_RENDITION_LIST` as it is.

File: src/js/experimental/media-rendition-listbox.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    super.disconnectedCallback();
+  }
+
+  get mediaRenditionList() {
+    return this.#renditionList;
+  }
+
+  set mediaRenditionList(list) {
+    this.removeAttribute(MediaUIAttributes.MEDIA_RENDITION_LIST);
```
Comment (heff) [inline src/js/experimental/media-rendition-listbox.js:76]: Why remove the attr here?

File: src/js/experimental/media-rendition-listbox.js [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+    this.#renditionList = list;
+    this.#render();
+  }
+
+  get mediaRenditionEnabled() {
+    if (this.value) {
+      return this.mediaRenditionList.filter(({ id }) => id == this.value);
```
Comment (heff) [inline src/js/experimental/media-rendition-listbox.js:84]: This can only ever return one item right? So no need to return an array?

File: src/js/experimental/media-rendition-listbox.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    return undefined;
```
Comment (heff) [inline src/js/experimental/media-rendition-listbox.js:86]: I'd prefer null here. I forget where @cjpillsbury and I landed on arrays specifically.

File: src/js/experimental/media-rendition-listbox.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  set mediaRenditionEnabled(list) {
+    this.removeAttribute(MediaUIAttributes.MEDIA_RENDITION_ENABLED);
```
Comment (heff) [inline src/js/experimental/media-rendition-listbox.js:90]: Same question as above on removing the attr.

File: src/js/experimental/media-chrome-listbox.js [hunk 7]
```diff
@@ -5,42 +5,39 @@ const template = document.createElement('template');
...
-  media-chrome-listitem[aria-selected="true"] {
+  ::slotted(media-chrome-option[aria-selected="true"]),
+  media-chrome-option[aria-selected="true"] {
     background-color: var(--media-listbox-selected-background, rgb(122 122 184 / .8));
   }
 
-  ::slotted(media-chrome-listitem:hover),
-  media-chrome-listitem:hover {
+  ::slotted(media-chrome-option:hover),
+  media-chrome-option:hover {
```
Comment (heff) [inline src/js/experimental/media-chrome-listbox.js:35]: Right now hovering on a selected item triggers the hover state, but it shouldn't. We should ignore this when selected.

File: src/js/controller.js (hunk 1 above)
Comment (heff) [reply src/js/controller.js:821]: If this is disabling all other renditions at the same time, so only one can be enabled at a time, then I'm really not sure why we're working with arrays for _enabled_.

File: src/js/media-container.js [hunk 8]
```diff
@@ -351,9 +351,13 @@ class MediaContainer extends window.HTMLElement {
   }
 
   static get observedAttributes() {
-    return [Attributes.AUTOHIDE, Attributes.GESTURES_DISABLED].concat(
-      MEDIA_UI_ATTRIBUTE_NAMES
-    );
+    return [Attributes.AUTOHIDE, Attributes.GESTURES_DISABLED]
+      .concat(MEDIA_UI_ATTRIBUTE_NAMES)
+      .filter(name => ![
```
Comment (heff) [inline src/js/media-container.js:356]: Why? (could at least use a comment)

File: src/js/media-controller.js [hunk 9]
```diff
@@ -219,39 +225,71 @@ class MediaController extends MediaContainer {
...
+    this.#handleVideoTrack({ type: 'removeall' });
+
     // Reset to paused state
     // TODO: Can we just reset all state here?
     // Should hasPlayed refer to the media element or the controller?
     // i.e. the poster might re-show if not handled by the poster el
     this.propagateMediaState(MediaUIProps.MEDIA_PAUSED, true);
   }
 
+  #handleVideoTrack = ({ type = '', track = null } = {}) => {
```
Comment (heff) [inline src/js/media-controller.js:259]: This is complicated. It'd be nice if we could pull this into the above logic more, though I see how it's different because there needs to be a listener on each individual track.

File: src/js/media-controller.js [hunk 10]
```diff
@@ -625,6 +666,18 @@ const propagateMediaState = (els, stateName, val) => {
   });
 };
 
+const getStateValue = (els, stateName) => {
```
Comment (heff) [inline src/js/media-controller.js:669]: Why was this needed?

**Review by heff — COMMENTED** [review]
Looks like it solves the need functionally. 👍 

The rendition data structures feel like they could use another pass. There seems to be competing designs happening between media-ui-extensions and what's being implemented here, like returning arrays that can only ever have one value and not actually using RENDITION_LIST anywhere.

File: src/js/constants.js [hunk 11]
```diff
@@ -57,6 +58,9 @@ export const MediaUIProps = {
   MEDIA_STREAM_TYPE: 'mediaStreamType',
   MEDIA_TARGET_LIVE_WINDOW: 'mediaTargetLiveWindow',
   MEDIA_TIME_IS_LIVE: 'mediaTimeIsLive',
+  MEDIA_RENDITION_LIST: 'mediaRenditionList',
+  MEDIA_RENDITION_ENABLED: 'mediaRenditionEnabled',
+  MEDIA_RENDITION_ACTIVE: 'mediaRenditionActive',
```
[the remark this answers is not in the document]
Comment (luwes) [reply src/js/constants.js:63]: yes, when the rendition is enabled / disabled it's not immediately playing because a stream switch can take a bit.
when the rendition starts playing then it becomes active...

File: src/js/controller.js [hunk 12]
```diff
@@ -470,13 +470,55 @@ export const MediaUIStates = {
...
+      const selected = media?.videoTracks?.[media.videoTracks?.selectedIndex ?? 0];
+
+      if (!selected) return [];
+
+      return [...selected.renditions ?? []];
+    },
+    mediaEvents: ['loadstart'],
+    renditionListEvents: ['addrendition', 'removerendition'],
+  },
+  MEDIA_RENDITION_ENABLED: {
```
[the remark this answers is not in the document]
Comment (luwes) [reply src/js/controller.js:487]: good call, will change to "renditions".

The rendition API supports enabling just a few renditions. Maybe only 270p and 360p for whatever reason. 
This would allow that. 

It's also one of the reasons why the `selectedIndex` property for renditions in media ui ext doesn't make sense.

File: src/js/controller.js [hunk 13]
```diff
@@ -470,13 +470,55 @@ export const MediaUIStates = {
...
+
+      // If all renditions are enabled it means `auto` is selected.
+      if (renditions.every(r => r.enabled)) return [];
+
+      return renditions.filter(r => r.enabled);
+    },
+    mediaEvents: ['loadstart'],
+    renditionListEvents: ['addrendition', 'removerendition', 'change'],
+  },
+  MEDIA_RENDITION_ACTIVE: {
```
[the remark this answers is not in the document]
Comment (luwes) [reply src/js/controller.js:504]: it's using arrays in the same fashion as `MEDIA_SUBTITLES_LIST` and `MEDIA_SUBTITLES_SHOWING`.
it makes it easier to use the stringify and parse utilities for all of them.

File: src/js/controller.js (hunk 1 above)
Comment (luwes) [reply src/js/controller.js:821]: the renditions API supports enabling 1, multiple or all renditions so an array makes sense.

via the UI we'll probably not support the multiple enable though...

File: src/js/experimental/media-rendition-listbox.js (hunk 3 above)
Comment (luwes) [reply src/js/experimental/media-rendition-listbox.js:76]: this is related to @cjpillsbury feedback of making some rendition related API only via props (not reflecting to attrs) but still accept attributes.

the prop / attribute is not reflected anymore so we used this behavior for mediaController.fullscreenElement in removing the attribute if the property was used.

the alternative is if there was ever an attribute set to also keep updating the attribute if ever the property was set but would make it more complex I think.

File: src/js/media-controller.js (hunk 10 above)
Comment (luwes) [reply src/js/media-controller.js:669]: before we could easily check all the attribute values on media-controller but since we will not set some complex data attributes like mediarenditionlist on <media-controller> we'll have to get the values from one of the control media state receivers. cc @cjpillsbury

File: src/js/media-controller.js (hunk 9 above)
Comment (luwes) [reply src/js/media-controller.js:259]: yes, that's what I was hinting at in point 1 in the PR description.
I don't like it that much either.

File: src/js/experimental/media-rendition-button.js (hunk 2 above)
Comment (luwes) [reply src/js/experimental/media-rendition-button.js:48]: `MEDIA_RENDITION_LIST` is used in rendition listbox, it still accepts the attribute but it'll prefer to use the property if it is available

File: src/js/experimental/media-rendition-listbox.js (hunk 4 above)
Comment (luwes) [reply src/js/experimental/media-rendition-listbox.js:84]: this is solved with making this plural mediaRenditionsEnabled because there can be a case w/ multiple renditions enabled

File: src/js/experimental/media-rendition-listbox.js (hunk 5 above)
Comment (luwes) [reply src/js/experimental/media-rendition-listbox.js:86]: I'm fine with changing. we use `undefined` in more places for props currently I think (e.g. in Mux player) but null is used a lot in DOM API's, getAttribute returns null too, className,

File: src/js/experimental/media-rendition-listbox.js (hunk 3 above)
Comment (heff) [reply src/js/experimental/media-rendition-listbox.js:76]: Yeah, that makes sense. Complex values shouldn't be sprouted because you're not going to style things with them, and even simplifying the complex values would open up other complexity.

Deleting it is still form of sprouting, but I guess things like `autoplay` already do that.

File: src/js/experimental/media-rendition-listbox.js (hunk 5 above)
Comment (heff) [reply src/js/experimental/media-rendition-listbox.js:86]: Yeah, let's use null here. And in general save `undefined` specifically for unimplemented APIs and functions that are not intended to ever return anything. That's my experience of how most JS APIs are in HTML.

File: src/js/media-controller.js (hunk 10 above)
Comment (heff) [reply src/js/media-controller.js:669]: Probably fine but feels a little odd to use a list of stateReceivers as the source of truth for the prior state. I wonder of we could use MediaUIStates, here or if that'd already be changed at this point. This might be a reason to move forward with state properties for media controller specifically.

File: src/js/experimental/media-rendition-listbox.js (hunk 3 above)
Comment (luwes) [reply src/js/experimental/media-rendition-listbox.js:76]: yes, I can see that. do you think it's better to leave it as is? not remove the attribute.
I'm fine with both I think, it's a pretty rare case

File: src/js/controller.js [hunk 14]
```diff
@@ -470,13 +470,29 @@ export const MediaUIStates = {
...
+    },
+    mediaEvents: ['emptied'],
+    videoRenditionsEvents: ['addrendition', 'removerendition'],
+  },
+  MEDIA_RENDITION_SELECTED: {
+    get: function (controller) {
+      const { media } = controller;
+      return media?.videoRenditions?.[media.videoRenditions?.selectedIndex];
+    },
+    mediaEvents: ['emptied'],
```
Comment (heff) [inline src/js/controller.js:488]: Is there no event we should be listening to for the first load? I worry a little about missing an add rendition event, if that's what's being relied on.

File: src/js/controller.js [hunk 15]
```diff
@@ -470,13 +470,29 @@ export const MediaUIStates = {
...
+      const { media } = controller;
+      return [...media?.videoRenditions ?? []];
+    },
+    mediaEvents: ['emptied'],
+    videoRenditionsEvents: ['addrendition', 'removerendition'],
+  },
+  MEDIA_RENDITION_SELECTED: {
+    get: function (controller) {
+      const { media } = controller;
+      return media?.videoRenditions?.[media.videoRenditions?.selectedIndex];
```
Comment (heff) [inline src/js/controller.js:486]: Nice, the API turned out relatively simple, even without IDs.

I think this property should just return the index value rather than the whole rendition object. As is it's duplicate info with the list, and using just the index means it could still be a simple HTML attribute, even if the list is more complex.

File: src/js/controller.js [hunk 16]
```diff
@@ -769,4 +785,12 @@ export const MediaUIRequestHandlers = {
 
     media.currentTime = seekable.end(seekable.length - 1);
   },
+  MEDIA_RENDITION_REQUEST: (media, event) => {
+    if (!media?.videoRenditions) return;
```
Comment (heff) [inline src/js/controller.js:789]: Let's add a console warning here. i.e. "Rendition selection not supported by this media". Silent failing here could lead to some frustration.

File: src/js/experimental/media-rendition-button.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+  <path d="M13.5 2.5h2v6h-2v-2h-11v-2h11v-2Zm4 2h4v2h-4v-2Zm-12 4h2v6h-2v-2h-3v-2h3v-2Zm4 2h12v2h-12v-2Zm1 4h2v6h-2v-2h-8v-2h8v-2Zm4 2h7v2h-7v-2Z" />
+</svg>`;
+
+const slotTemplate = document.createElement('template');
+slotTemplate.innerHTML = /*html*/`
+  <slot name="icon">${renditionIcon}</slot>
+`;
+
+/**
+ * @attr {string} mediarenditionsenabled - (read-only) Set to the enabled rendition.
```
Comment (heff) [inline src/js/experimental/media-rendition-button.js:38]: "enabled" is wrong now, right?

File: src/js/experimental/media-rendition-button.js [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+  constructor() {
+    super({ slotTemplate });
+  }
+
+  get mediaRenditionSelected() {
+    return getRenditionListAttr(this, MediaUIAttributes.MEDIA_RENDITION_SELECTED)[0];
+  }
+
+  set mediaRenditionSelected(rendition) {
+    const value = rendition ? [rendition] : null;
```
Comment (heff) [inline src/js/experimental/media-rendition-button.js:59]: It doesn't need to be an array anymore, right? Or why is this an array?

File: src/js/experimental/media-rendition-listbox.js [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+      container.append(this.#autoOption);
+    }
+
+    let isAuto = !this.mediaRenditionSelected;
+    if (isAuto) {
+      this.#autoOption.setAttribute('aria-selected', 'true');
+      this.#autoOption.setAttribute('tabindex', '0');
+    } else {
+      this.#autoOption.setAttribute('aria-selected', 'false');
+      this.#autoOption.setAttribute('tabindex', '-1');
```
Comment (heff) [inline src/js/experimental/media-rendition-listbox.js:113]: Why is auto getting a -1 tabindex here? I'm assuming we still want auto tab selectable? Probably worth a comment.

File: src/js/experimental/media-rendition-listbox.js [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+
+    for (const rendition of renditionList) {
+
+      /** @type {HTMLOptionElement} */
+      const option = document.createElement('media-chrome-option');
+      option.part.add('option');
+      option.value = `${rendition.id}`;
+      option.textContent = `${Math.min(rendition.width, rendition.height)}p`;
+
+      if (rendition.enabled && !isAuto) {
```
Comment (heff) [inline src/js/experimental/media-rendition-listbox.js:124]: Enabled's not a thing anymore, right?

File: src/js/media-container.js [hunk 21]
```diff
@@ -248,6 +248,17 @@ function getBreakpoints(breakpoints, rect) {
  * @cssprop --media-slot-display - `display` of the media slot (default none for [audio] usage).
  */
 class MediaContainer extends globalThis.HTMLElement {
+  static get observedAttributes() {
+    return [Attributes.AUTOHIDE, Attributes.GESTURES_DISABLED]
+      .concat(MEDIA_UI_ATTRIBUTE_NAMES)
+      // Filter out specific / complex data media UI attributes
+      // that shouldn't be propagated to this state receiver element.
```
Comment (heff) [inline src/js/media-container.js:255]: Does it break somehow? Is the list wrong if these aren't actually attributes?

**Review by heff — COMMENTED** [review]
Looking a lot better and pretty close! The main design question is about the selected rendition being a list rather than just the index value.

File: src/js/controller.js (hunk 14 above)
Comment (luwes) [reply src/js/controller.js:488]: I can see that. 
let's add `loadstart` as well

File: src/js/controller.js (hunk 15 above)
Comment (luwes) [reply src/js/controller.js:486]: after implementing this I remembered why I'd prefer also the complex object here. 
because often (and I would like to use) the quality value to style the rendition / quality button. 
see screenshot

SCR-20230725-mod


by adding the width and height of the selected rendition we can add HD, 4K labels or icons easily to the button with CSS...

File: src/js/experimental/media-rendition-listbox.js (hunk 19 above)
Comment (luwes) [reply src/js/experimental/media-rendition-listbox.js]: I removed this from this PR. I'll address it in the other one.
I think it is needed once you select an option, it's how it's done.
Set the selected option to tabindex=0 and the rest of the options to tabindex=-1

File: src/js/experimental/media-rendition-listbox.js (hunk 20 above)
Comment (luwes) [reply src/js/experimental/media-rendition-listbox.js]: good catch! this is selected

File: src/js/media-container.js (hunk 21 above)
Comment (luwes) [reply src/js/media-container.js:255]: these are filtered out because they can get pretty big.
but no, it would not break anything.
they just take a lot of HTML space so to speak.

File: src/js/controller.js (hunk 15 above)
Comment (heff) [reply src/js/controller.js:486]: To the button itself, not the options? What would that look like exactly? And can the CSS be used reliably with varying widths and heights?

Comment (luwes) [reply src/js/controller.js:486]: good points, I also just reminded it would work better with also having the quality that is actually playing.

something like this

SCR-20230725-qod

Comment (luwes) [reply src/js/controller.js]: going with rendition index only the `mediarenditionselected` attribute

File: src/js/media-container.js (hunk 21 above)
Comment (luwes) [reply src/js/media-container.js:255]: only the attribute `MediaUIAttributes.MEDIA_RENDITION_LIST` is filtered out now.
it's set directly via a JS property w/ fallback to attribute

**Review by heff — APPROVED** [review]
Looks good!
